### ISSUE TITLE
Session [*] was terminated due to SO_TIMEOUT

### ISSUE THREAD

**Issue by benzman81** [SeleniumHQ/selenium#2400]
## Meta -
OS:  
Jenkins running in Linux usind TestNG to run Tests in Selenium Grid where the nodes run in Windows 7
Selenium Version: 
WebDriver: 2.48.2
IEDriver: 2.48.0
Browser:  
Internet Explorer
Browser Version:  
11

## Expected Behavior -
Tests run Multithreaded using TestNG on multiple nodes without timeour.

## Actual Behavior -
Our Tests run only 2 out of about 8 times a day without an SO_TIMEOUT. The other test runs, suddenly the tests start failing with the following stacktrace:

`
    Exception org.openqa.selenium.WebDriverException
        
    Message: Session [6919ba25-53b6-4615-bd59-e97399bf1e12] was terminated due to SO_TIMEOUT Command duration or timeout: 3 milliseconds Build info: version: '2.47.1', revision: 'unknown', time: '2015-07-30 11:02:44' System info: host: 'myhost', ip: 'myip', os.name: 'Linux', os.arch: 'amd64', os.version: '3.2.0-23-generic', java.version: '1.7.0_05' Session ID: 6919ba25-53b6-4615-bd59-e97399bf1e12 Driver info: org.openqa.selenium.remote.RemoteWebDriver Capabilities [{platform=WINDOWS, javascriptEnabled=true, elementScrollBehavior=0, enablePersistentHover=true, ignoreZoomSetting=false, ie.ensureCleanSession=false, enableElementCacheCleanup=true, browserName=internet explorer, unexpectedAlertBehaviour=dismiss, webdriver.remote.sessionid=6919ba25-53b6-4615-bd59-e97399bf1e12, pageLoadStrategy=normal, version=11, ie.usePerProcessProxy=false, cssSelectorsEnabled=true, ignoreProtectedModeSettings=false, requireWindowFocus=false, initialBrowserUrl=http://localhost:12863/, handlesAlerts=true, ie.forceCreateProcessApi=false, nativeEvents=true, browserAttachTimeout=0, ie.browserCommandLineSwitches=, takesScreenshot=true}]
    Stacktrace:
        at sun.reflect.GeneratedConstructorAccessor83.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:525)
        at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
        at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:595)
        at org.openqa.selenium.remote.RemoteWebDriver.get(RemoteWebDriver.java:306)
        .
        .
        .
        some trace from our code that calls RemoteWebDriver.get(url)
        .
        .
        .
        at sun.reflect.GeneratedMethodAccessor24.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:601)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
        at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:510)
        at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:211)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:585)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:816)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1124)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
        at java.lang.Thread.run(Thread.java:722)
    Caused by: org.openqa.grid.common.exception.GridException: Session [6919ba25-53b6-4615-bd59-e97399bf1e12] was terminated due to SO_TIMEOUT
        at org.openqa.grid.internal.ActiveTestSessions.getExistingSession(ActiveTestSessions.java:105)
        at org.openqa.grid.internal.Registry.getExistingSession(Registry.java:425)
        at org.openqa.grid.web.servlet.handler.RequestHandler.getSession(RequestHandler.java:240)
        at org.openqa.grid.web.servlet.handler.RequestHandler.process(RequestHandler.java:120)
        at org.openqa.grid.web.servlet.DriverServlet.process(DriverServlet.java:83)
        at org.openqa.grid.web.servlet.DriverServlet.doPost(DriverServlet.java:67)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.seleniumhq.jetty9.servlet.ServletHolder.handle(ServletHolder.java:808)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doHandle(ServletHandler.java:587)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doHandle(SessionHandler.java:221)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doScope(ServletHandler.java:515)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doScope(SessionHandler.java:185)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.seleniumhq.jetty9.server.Server.handle(Server.java:499)
        at org.seleniumhq.jetty9.server.HttpChannel.handle(HttpChannel.java:310)
        at org.seleniumhq.jetty9.server.HttpConnection.onFillable(HttpConnection.java:257)
        at org.seleniumhq.jetty9.io.AbstractConnection$2.run(AbstractConnection.java:540)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.lang.Thread.run(Thread.java:745)
`

We have not clue anymore what the problem might be. Today be upgrade WebDriver and IEDriver to the latest releases, but going through the changelog doesnt semm that this is fixed.

Out timeouts are set like this:
`
    "nodePolling": 5000,
    "cleanUpCycle": 5000,
    "browserTimeout": 300000,
    "timeout": 600000,
`

Any ideas whats going on or if we need some different timeout settings?

## Steps to reproduce -
Not found a way to reproduce it although this happens more than once a day.

**Comment by p00j4** [comment]
duplicate of https://github.com/seleniumhq/selenium-google-code-issue-archive/issues/4566
@benzman81 try with proper timeout settings as per this discussion and suggestions by lukeis

**Comment by benzman81** [comment]
@p00j4 In this issue its more about "TIMEOUT" and "BROWSER_TIMEOUT". I would change the settings, if I would know what a "proper" timeout setting woulds be. Especially for avoiding a "SO_TIMEOUT" as I could not find a setting for this. We have about 500 tests that run in parallel using 10 threads.If the "SO_TIMEOUT" exception occurs once, than all following tests will fail with this error. So does not seem to me that this is an exception based on a single test. If we start a new run of these tests after failing in previous run with the exception, than most of the time it works again until the situation happens again. Maybe someone could explain in what cases a "SO_TIMEOUT" exception might occur and what timeout setting is repsonsible for this.

**Comment by Jan0707** [comment]
Facing the same random problem here. Tried to check the documentation on what "SO_TIMEOUT" actually is - no luck so far. Any progress ?

**Comment by lukeis** [comment]
SO_TIMEOUT means Socket Timeout... aka it didn't get a response within the specified timeout for the HTTP request.

**Comment by benzman81** [comment]
Do you know how we can change this timeout. Did not find a setting for socket timeout.

**Comment by lukeis** [comment]
your exact timeout that you want configured? You'll have to figure that bit out... but I do feel the help text on the configuration is better in the 3.0 beta release:

for 'node' parameters:
java -jar selenium-server-standalone-3.0.0-beta2.jar -role node -h

for 'hub' parameters:
java -jar selenium-server-standalone-3.0.0-beta2.jar -role hub -h

Since i don't see anything in this issue that appears to be a bug in the grid, rather just configuration needs tweaking, going to close this issue. If you can provide a reproducible set of actions, please let us know.

**Comment by benzman81** [comment]
Well, as stated in a comment above, I just need to know which configuration parameter is responsible, since it is not stated anywhere for SO_TIMEOUTS. If I would know which parameter to tweak for the SO_TIMEOUTS I would not have opened an issue here. At least it should be listed in a documentation or FAQ or somewhere else what to tweak for SO_TIMEOUTS.

**Comment by lukeis** [comment]
I suggest following up in the selenium-users google group:
https://groups.google.com/forum/#!forum/selenium-users

Pull requests happily accepted to any changes to the documentation

**Comment by yurygnutov** [comment]
@benzman81 hello. Did you finally reach the point of solve the problem? We are experiencing the same issue.

**Comment by benzman81** [comment]
@yurygnutov since updating to latest selenium 3.x with driver nad grid the issue doesnt happen anymore. There was also an issue where the Internetexplorer was eating to much memory and so not responding anymore. I can't quite remember if that helped, too.

**Comment by kurtfusion** [comment]
I ran into a issues similar today. This SO_TIMEOUTS happened when the hub was running 2.53.1 but was configured with the 3.6 hub configuration file.

**Comment by dizeee** [comment]
Happens from time to time with 3.8.1.

**Comment by ekrotenko** [comment]
Happens unexpectedly with with 3.8.1
